File: Modulo 4 - Tipos Parametricos/1/1.1.c
#include<stdio.h>

// Cada personagem tem Altura e Idade
typedef struct Base {
	float altura;
	int idade;
} Base;


// O personagem e:
// Um Guerreiro que tem forca
// Um Mago que tem magia
// Um Sacerdote que tem reza
enum CLASSE {Guerreiro, Mago, Sacerdote};

typedef struct Personagem {
	enum CLASSE sub;
	Base status;
	union {
		struct { int forca; }; // Guerreiro
		struct { float magia; }; // Mago
		struct { int reza; }; // Sacerdote
	};
} Personagem;

int main () {
	Personagem p1 = { Guerreiro, {1.8, 23}, {.forca=50}};
	Personagem p2 = { Mago, {1.6, 46}, {.magia=80}};
	
	printf("Guerreiro - Altura: %f - Idade: %d - Forca: %d", p1.status.altura, p1.status.idade, p1.forca);
	printf("\n");
	printf("Mago - Altura: %f - Idade: %d - Magia: %f", p2.status.altura, p2.status.idade, p2.magia);
	
	return 0;
}

File: Modulo 4 - Tipos Parametricos/1/1.1.hs
-- Cada personagem tem Altura e Idade
data Base = Base Double Int
    deriving Show

-- O personagem e:
-- Um Guerreiro que tem forca
-- Um Mago que tem magia
-- Um Sacerdote que tem reza
data Personagem = Guerreiro Base Int | Mago Base Double | Sacerdote Base Int
    deriving Show
 

p1 :: Personagem
p1 = Guerreiro (Base 1.8 23) 50

p2 :: Personagem
p2 = Mago (Base 1.6 46) 80

main = print(p1," - ",p2)